homework-submission/week5-class16/src/stores/TodosStore.js

```diff
@@ -0,0 +1,174 @@
+import { observable, computed, action, runInAction } from "mobx";
+
+import { configure } from "mobx";
+configure({ enforceActions: "observed" });
+
+const API_Base = "https://hyf-react-api.herokuapp.com";
+
+class TodosStore {
+  @observable
+  edit = false;
+  @observable
+  Data = [];
+  @observable
+  state = "loading";
+
+  @action
+  handleCheck = async todoId => {
+    this.state = "loading";
+    const targetedTodo = this.Data.find(todo => todo._id === todoId);
+    // this code block to show instant change to the user.
+    const newData = this.Data.map(todo => {
+      if (todo._id === todoId) {
+        return { ...todo, done: !todo.done };
+      }
+      return todo;
+    });
+    this.Data = newData;
+    this.state = "done";
+
+    try {
+      await this.patchRequest(todoId, { done: !targetedTodo.done });
+    } catch (error) {
+      runInAction(() => {
+        this.state = "error";
+      });
+    }
+
+    this.getTodos();
+  };
+  @action
+  handleDelete = async todoId => {
+    this.state = "loading";
+
+    try {
+      await this.deleteRequest(todoId);
+      runInAction(() => {
+        this.state = "done";
+      });
+    } catch (error) {
+      runInAction(() => {
+        this.state = "error";
+      });
+    }
+
+    this.getTodos();
+  };
+
+  @action
+  handleAddTodo = async (description, deadline) => {
+    this.state = "loading";
+    const newTodo = {
+      description,
+      deadline,
+      done: "false"
+    };
+    try {
+      await this.postRequest(newTodo);
+      runInAction(() => {
+        this.state = "done";
+      });
+    } catch (error) {
+      runInAction(() => {
+        this.state = "error";
+      });
+    }
+    this.getTodos();
+  };
+
+  @action
+  handleEdit = todoId => {
+    const toEdit = this.Data.find(todo => todo._id === todoId);
+    toEdit.edit = !this.edit;
+  };
+  @action
+  cancelEdit = todoId => {
+    const toCancel = this.Data.find(todo => todo._id === todoId);
+    toCancel.edit = !toCancel.edit;
+  };
+
+  @action
+  handleUpdate = async (todoId, description, deadline) => {
+    this.state = "loading";
+
+    let targetTodo = this.Data.find(todo => todo._id === todoId);
+    const updatedTodo = {
+      ...targetTodo,
+      description: description || targetTodo.description,
+      deadline: deadline || targetTodo.deadline
+    };
+    delete updatedTodo.edit;
```

Comment:
Nice to delete it before send for a `bit` faster network request, I am not sure about this but I think most servers if they are not expecting it they will just ignore it.